src/app/cluster/cluster-details/constraints/constraint-dialog/component.ts

```diff
@@ -0,0 +1,142 @@
+  templateUrl: './template.html',
+  styleUrls: ['./style.scss'],
+})
+export class ConstraintDialog implements OnInit, OnDestroy {
+  readonly Controls = Controls;
+  form: FormGroup;
+  spec = '';
+  constraintTemplates: ConstraintTemplate[] = [];
+  private readonly _unsubscribe = new Subject<void>();
+
+  constructor(
+    private readonly _matDialogRef: MatDialogRef<ConstraintDialog>,
+    private readonly _opaService: OPAService,
+    private readonly _notificationService: NotificationService,
+    private readonly _builder: FormBuilder,
+    @Inject(MAT_DIALOG_DATA) public data: ConstraintDialogConfig
+  ) {}
+
+  ngOnInit(): void {
+    this.form = this._builder.group({
+      [Controls.Name]: this._builder.control(this.data.mode === Mode.Edit ? this.data.constraint.name : '', [
+        Validators.required,
+      ]),
+      [Controls.ConstraintTemplate]: this._builder.control(
+        this.data.mode === Mode.Edit ? this.data.constraint.spec.constraintType : '',
+        [Validators.required]
+      ),
+    });
+
+    this._initConstraintSpecEditor();
+
+    this._opaService.constraintTemplates
+      .pipe(takeUntil(this._unsubscribe))
+      .subscribe(constraintTemplates => (this.constraintTemplates = constraintTemplates));
+  }
+
+  ngOnDestroy(): void {
+    this._unsubscribe.next();
+    this._unsubscribe.complete();
+  }
+
+  save(): void {
+    const constraint: Constraint = {
+      name: this.form.get(Controls.Name).value,
+      spec: this._getSpec(),
+    };
+
+    switch (this.data.mode) {
+      case Mode.Add:
+        return this._create(constraint);
+      case Mode.Edit:
+        return this._edit(constraint);
+    }
+  }
+
+  private _create(constraint: Constraint): void {
+    this._opaService
+      .createConstraint(this.data.projectId, this.data.clusterId, constraint)
+      .pipe(take(1))
+      .subscribe(result => {
+        this._matDialogRef.close(true);
+        this._notificationService.success(`The constraint ${result.name} was created`);
+        this._opaService.refreshConstraint();
+      });
+  }
+
+  private _edit(constraint: Constraint): void {
+    this._opaService
+      .patchConstraint(this.data.projectId, this.data.clusterId, this.data.constraint.name, constraint)
+      .pipe(take(1))
+      .subscribe(result => {
+        this._matDialogRef.close(true);
+        this._notificationService.success(`The constraint ${result.name} was updated`);
+        this._opaService.refreshConstraint();
+      });
+  }
+
+  private _initConstraintSpecEditor(): void {
+    if (this.data.mode === Mode.Edit) {
+      const spec = this.data.constraint.spec;
+      if (!_.isEmpty(spec)) {
+        this.spec = dump(spec);
+      }
+    }
+  }
+
+  private _getSpec(): ConstraintSpec {
+    let spec = new ConstraintSpec();
+    const raw = load(this.spec) as ConstraintSpec;
+    if (!_.isEmpty(raw)) {
+      spec = raw;
+    }
+    spec['constraintType'] = this.form.get(Controls.ConstraintTemplate).value;
```

Comment:
Looks better. Since `spec` is now of `ConstraintSpec` type there is no need to use bracket notation to get the value. Take it from the object instead.